## Numpy Tutorial

NumPy (Numerical Python) is a core Python library used for efficient numerical computation. It provides a powerful data structure called the **ndarray (N-dimensional array)**, which allows storing and manipulating large amounts of data in a compact and structured way. Unlike Python lists, NumPy arrays are homogeneous, meaning all elements have the same data type, which makes them more memory-efficient and suitable for mathematical operations.

The main concept behind NumPy is **vectorization**, which enables operations to be performed on entire arrays at once instead of using loops. This leads to cleaner code and significantly faster execution. NumPy also supports broadcasting, allowing operations between arrays of different shapes without explicitly reshaping them.

NumPy is very fast due to several key reasons. First, it is implemented in low-level languages like C, so most operations run outside the Python interpreter, reducing overhead. Second, it uses contiguous memory storage, which improves cache efficiency and speeds up data access. Third, vectorized operations eliminate the need for Python loops, which are slower compared to compiled code execution. Additionally, NumPy performs optimized numerical computations using highly efficient algorithms.

It also provides a wide range of built-in functions for linear algebra, statistical analysis, random number generation, and array manipulation. Because of its speed, efficiency, and powerful features, NumPy serves as the foundation for many advanced libraries used in data science, machine learning, and scientific computing.



---

# 📘 1. NumPy (Numerical Python)

### ✅ Definition

**NumPy** is a powerful Python library used for:

* Numerical computations
* Handling large arrays and matrices
* Performing mathematical operations efficiently

### ✅ Key Features

* Fast (written in C)
* Supports multi-dimensional arrays
* Vectorized operations (no need for loops)
* Rich mathematical functions

### ✅ Example

```python
import numpy as np

arr = np.array([1, 2, 3, 4])
print(arr)
```

**Output:**

```
[1 2 3 4]
```

👉 Meaning: Creates a NumPy array from a Python list.

---

# 📘 2. ndarray (N-dimensional Array)

### ✅ Definition

**ndarray** is the core data structure of NumPy.

* Stores elements of the **same data type**
* Supports **1D, 2D, 3D, ... nD arrays**

### ✅ Syntax

```python
np.array(object, dtype=None, ndmin=0)
```

### ✅ Parameters

* `object` → list/tuple/array input
* `dtype` → data type (int, float, etc.)
* `ndmin` → minimum number of dimensions

---

### ✅ Example 1 (1D Array)

```python
arr = np.array([10, 20, 30])
print(arr)
```

### ✅ Example 2 (2D Array)

```python
arr = np.array([[1, 2], [3, 4]])
print(arr)
```

### ✅ Example 3 (dtype)

```python
arr = np.array([1, 2, 3], dtype=float)
print(arr)
```

👉 Meaning: Converts all values to float.

---

### ✅ Important Attributes

```python
arr.shape   # dimensions
arr.ndim    # number of axes
arr.size    # total elements
arr.dtype   # data type
```

---

# 📘 3. Advantages of NumPy

### ✅ 1. High Performance

* Faster than Python lists (C backend)

### ✅ 2. Memory Efficient

* Uses less memory due to fixed data type

### ✅ 3. Vectorization

```python
arr = np.array([1,2,3])
print(arr * 2)
```

👉 No loop needed → faster computation

---

### ✅ 4. Supports Mathematical Operations

```python
np.mean(arr)
np.sum(arr)
np.sqrt(arr)
```

---

### ✅ 5. Broadcasting
Broadcasting in NumPy is a powerful feature that allows arithmetic operations between arrays of different shapes without explicitly reshaping them.
NumPy automatically "broadcasts" the smaller array across the larger one so that they have compatible shapes for element-wise operations.
```python
arr = np.array([1,2,3])
print(arr + 5)
```

👉 Adds 5 to all elements

---

### ✅ 6. Multi-dimensional Support

* Useful in ML, Data Science

---

# 📘 4. Memory Model of NumPy

### ✅ Key Idea

NumPy arrays store data in **contiguous memory blocks**

👉 Unlike Python lists (which store references)

---

### ✅ Python List Memory

* Stores pointers to objects
* Different memory locations

```
[1, 2, 3]
→ pointer → 1
→ pointer → 2
→ pointer → 3
```

---

### ✅ NumPy Array Memory

* Stores actual values in **continuous memory**

```
[1 2 3] → stored in one block
```

---

### ✅ Benefits

* Faster access
* Better cache performance
* Less memory usage

---

### ✅ Example (View vs Copy)

```python
arr = np.array([1,2,3])
b = arr

b[0] = 100
print(arr)
```

👉 Output:

```
[100 2 3]
```

✔ Same memory → change affects original

---

### ✅ Copy Example

```python
b = arr.copy()
```

👉 Now memory is separate

---

### ✅ Strides Concept (Important)

* Defines how many bytes to skip to move to next element

```python
arr.strides
```

**Strides** tell NumPy:

> “How many bytes to jump in memory to move to the next element along each axis”

👉 It is part of NumPy’s **memory model**

### ✅ Output

* Returns a tuple of bytes for each dimension

# 📘 Simple Example (1D Array)

```python
import numpy as np

arr = np.array([10, 20, 30], dtype=np.int32)
print(arr.strides)
```

### ✅ Output

```
(4,)
```

### 🔍 Meaning

* `int32` = 4 bytes
* To move from:

  * 10 → 20 → 30 → jump **4 bytes**

👉 So stride = `(4,)`

# 📘 Example (2D Array)

```python
arr = np.array([[1,2,3],
                [4,5,6]], dtype=np.int32)

print(arr.strides)
```

### ✅ Output

```
(12, 4)
```

---

### 🔍 Meaning

* Each row has 3 elements → 3 × 4 bytes = 12 bytes
* So:

  * Move to next **row** → jump 12 bytes
  * Move to next **column** → jump 4 bytes

👉 So:

```
(12, 4)
(row jump, column jump)
```

---

# 📘 Visual Understanding

Matrix:

```
[1  2  3
 4  5  6]
```

Memory (flattened):

```
[1][2][3][4][5][6]
```

* Horizontal move → +4 bytes
* Vertical move → +12 bytes

---

# 📘 Why Strides are Important?

### ✅ 1. Fast Computation

* Helps NumPy access memory efficiently

### ✅ 2. Views (No Copy)

```python
b = arr[:, ::2]
```

👉 Uses stride tricks → no new memory

---

### ✅ 3. Memory Optimization

* Avoid unnecessary copies
* Enables slicing efficiently

---

# 📘 Important Interview Point 🔥

👉 **Strides define how NumPy interprets memory layout**

Formula:

```
Memory Address = Base Address + Σ(index × stride)
```

---

# 📘 Quick Summary

* `arr.strides` → tuple of byte steps
* Depends on:

  * Data type (`dtype`)
  * Shape of array
* Used internally for:

  * Slicing
  * Views
  * Performance optimization

---

# 📘 5. Installation of NumPy

### ✅ Using pip

```bash
pip install numpy
```

### ✅ Using conda

```bash
conda install numpy
```

---

### ✅ Verify Installation

```python
import numpy as np
print(np.__version__)
```

---

### ✅ Upgrade

```bash
pip install --upgrade numpy
```

---

# 📘 Summary (Quick Revision)

* **NumPy** → Library for numerical computing
* **ndarray** → Core array object
* **Advantages** → Fast, memory efficient, vectorized
* **Memory Model** → Contiguous memory storage
* **Installation** → `pip install numpy`

---


In [8]:
import numpy as np

a = [1,2,3,4,5,6,7,8,9]
arr = np.array(a)
print(np.array(a))
# if we add floating or string elements it will auto convert all elements.
# dtype = means the datatype that can auto conv or accept that type only
# np.array(object, dtype=None, copy=True, order='K', subok=False, ndmin=0)

# Rule 
# int + float = Float
# int + string = String

[1 2 3 4 5 6 7 8 9]


In [9]:
arr.ndim   # dimensions

1

In [10]:
arr.shape  # structure


(9,)

In [11]:
arr.size   # total elements

9

In [12]:

arr.dtype  # data type

dtype('int64')

## Indexing and slicing

In [13]:
print(arr)

[1 2 3 4 5 6 7 8 9]


In [14]:
print(arr[1,2]) # this is only for multidimensionals
# if we try that we can get the error


IndexError: too many indices for array: array is 1-dimensional, but 2 were indexed

In [15]:
print(arr[0])

1


In [16]:
print(arr[1:4]) # just like list for 1D - array only


[2 3 4]


In [17]:
Two_d_array = np.array([[1,2,3],[4,5,6],[7,8,9]])
multi_array = np.array([[[10,20,30,40],[500,600,700,800],[2000,3000,6000,9000]],
                       [[19,290,380,456],[520,640,780,830],[2200,3300,6400,900]]])
print(Two_d_array)
# here down for slicing in 2D - array

[[1 2 3]
 [4 5 6]
 [7 8 9]]


In [19]:
# array[row, col]
print(Two_d_array[0,2]) # and it will differe when it comes to 3D - array

3


In [20]:
print(multi_array)

[[[  10   20   30   40]
  [ 500  600  700  800]
  [2000 3000 6000 9000]]

 [[  19  290  380  456]
  [ 520  640  780  830]
  [2200 3300 6400  900]]]


In [21]:
print(multi_array[1,2,2]) # array(no.of.arr, row, col)


6400


## Slicing

In [22]:
print(multi_array[1,1,2:])

[780 830]


### np.zeros & np.ones

In [23]:
# np.zeros(shape, dtype = float, order = 'c')
# Memory layout ('C' row-major, 'F' column-major)

np.zeros((2,5),dtype = np.int64)

array([[0, 0, 0, 0, 0],
       [0, 0, 0, 0, 0]])

In [24]:
# np.ones(shape,dtype=float,order='c'

np.ones((3,3),dtype = float)


array([[1., 1., 1.],
       [1., 1., 1.],
       [1., 1., 1.]])

In [25]:
np.ones(4,dtype = int)

array([1, 1, 1, 1])

### np.arange & np.linspace

In [26]:
# np.arange(start=0,stop,step = 1,dtype = None)
np.arange(5)

array([0, 1, 2, 3, 4])

In [27]:
np.arange(2,9,3)

array([2, 5, 8])

In [28]:
np.arange(2,9,0.7)


array([2. , 2.7, 3.4, 4.1, 4.8, 5.5, 6.2, 6.9, 7.6, 8.3])

In [29]:
# np.linspace(start,stop,num = 50,endpoint=True,retstep=False,dtype = None)
# num = 50 means by default no of values
# endpoint that includes stop or not
# retstep means return step size also

np.linspace(0,1,5,endpoint=True)

array([0.  , 0.25, 0.5 , 0.75, 1.  ])

In [30]:
np.linspace(0,1,5,endpoint=False)
# it will remove the endpoint so default is 0.1

array([0. , 0.2, 0.4, 0.6, 0.8])

In [31]:
np.linspace(1,10,endpoint = False)
# len(np.linspace(1,10,endpoint = False)) -> 50


array([1.  , 1.18, 1.36, 1.54, 1.72, 1.9 , 2.08, 2.26, 2.44, 2.62, 2.8 ,
       2.98, 3.16, 3.34, 3.52, 3.7 , 3.88, 4.06, 4.24, 4.42, 4.6 , 4.78,
       4.96, 5.14, 5.32, 5.5 , 5.68, 5.86, 6.04, 6.22, 6.4 , 6.58, 6.76,
       6.94, 7.12, 7.3 , 7.48, 7.66, 7.84, 8.02, 8.2 , 8.38, 8.56, 8.74,
       8.92, 9.1 , 9.28, 9.46, 9.64, 9.82])

In [32]:
np.linspace(1,20,3) # it will take 3 values with stop value 20

array([ 1. , 10.5, 20. ])

### np.zeros_like & np.ones_like

In [33]:
np.zeros_like(Two_d_array)

array([[0, 0, 0],
       [0, 0, 0],
       [0, 0, 0]])

In [34]:
np.ones_like(multi_array)

array([[[1, 1, 1, 1],
        [1, 1, 1, 1],
        [1, 1, 1, 1]],

       [[1, 1, 1, 1],
        [1, 1, 1, 1],
        [1, 1, 1, 1]]])

### np.full

In [35]:
# np.full(shape, fill_value, dtype = None)
np.full((4,4),"Hello", dtype = str)

array([['H', 'H', 'H', 'H'],
       ['H', 'H', 'H', 'H'],
       ['H', 'H', 'H', 'H'],
       ['H', 'H', 'H', 'H']], dtype='<U1')

In [36]:
np.full((3,3),100)

array([[100, 100, 100],
       [100, 100, 100],
       [100, 100, 100]])

### np.eye

`np.eye(N, M=None, k=0, dtype=float)`

- N -> rows
- m -> columns (optional)
- k -> diagonal index
    - 0 -> main diagonal
    - 1 -> above diagonal
    - -1 -> below diagonal

In [24]:
# np.eye(3) it will create the indenty matrix
np.eye(5,dtype = int) # by default k and dtype is default

array([[1, 0, 0, 0, 0],
       [0, 1, 0, 0, 0],
       [0, 0, 1, 0, 0],
       [0, 0, 0, 1, 0],
       [0, 0, 0, 0, 1]])

In [22]:
np.eye(4,6,k=1)

array([[0., 1., 0., 0., 0., 0.],
       [0., 0., 1., 0., 0., 0.],
       [0., 0., 0., 1., 0., 0.],
       [0., 0., 0., 0., 1., 0.]])

In [23]:
np.eye(4,6,k=-1)

array([[0., 0., 0., 0., 0., 0.],
       [1., 0., 0., 0., 0., 0.],
       [0., 1., 0., 0., 0., 0.],
       [0., 0., 1., 0., 0., 0.]])

#### np.indentity()
`np.identity(n,dtype,like)`

- it wil make the square identy matrix
-  identity is only for square
-  eye is for square and rectangular
-  identity has no futures like diagonal shift(k), columns control but eye does.



#### itemsize & nbytes

- itemsize tells you about the element that gonna takes the amount of space
- nbytes tells you about the total array size in bytes. no.of elements * 4 bytes.

In [51]:
import numpy as np
arr = np.identity(3)
print(arr)
a = np.array([1,2,3,4])
print(arr.itemsize) #itemsize (bytes) the element takes space in array
print(arr.nbytes)
print(a.nbytes) # complete array size in bytes. No.of elements * 4 bytes
# each integer takes 4 bytes


[[1. 0. 0.]
 [0. 1. 0.]
 [0. 0. 1.]]
8
72
32


| Feature    | `itemsize`              | `nbytes`                |
| ---------- | ----------------------- | ----------------------- |
| Meaning    | Memory per element      | Total memory of array   |
| Depends on | `dtype`                 | `size × itemsize`       |
| Value type | Small                   | Larger                  |
| Use        | Understand element size | Understand total memory |

#### flag table
| Flag           | Meaning                |
| -------------- | ---------------------- |
| `C_CONTIGUOUS` | Row-wise memory layout |
| `F_CONTIGUOUS` | Column-wise layout     |
| `OWNDATA`      | Array owns its memory  |
| `WRITEABLE`    | Can modify data        |



#### np.flags

`arr.flags` gives information about memory layout and properties of the array
- it tells, how data is stored in memory
- whether it is writable
- whether it owns its memory

In [53]:
a.flags

  C_CONTIGUOUS : True
  F_CONTIGUOUS : True
  OWNDATA : True
  WRITEABLE : True
  ALIGNED : True
  WRITEBACKIFCOPY : False



## ✅ Example

```python
import numpy as np

arr = np.array([[1,2,3],[4,5,6]])
print(arr.flags)
```

---

## ✅ Output (example)

```
  C_CONTIGUOUS : True
  F_CONTIGUOUS : False
  OWNDATA : True
  WRITEABLE : True
  ALIGNED : True
  WRITEBACKIFCOPY : False
```

---

# 📘 Important Flags Explained

---

## 🔹 1. `C_CONTIGUOUS`

### ✅ Meaning

* Data stored **row-wise (C-style)**
* Default in NumPy

👉 Example:

```
[1 2 3
 4 5 6]
```

Stored as → `[1,2,3,4,5,6]`

---

## 🔹 2. `F_CONTIGUOUS`

### ✅ Meaning

* Data stored **column-wise (Fortran-style)**

👉 Stored as → `[1,4,2,5,3,6]`

---

## 🔹 3. `OWNDATA`

### ✅ Meaning

* `True` → Array owns its memory
* `False` → It is a **view** of another array

---

## 🔹 4. `WRITEABLE`

### ✅ Meaning

* `True` → You can modify array
* `False` → Read-only array

---

### ✅ Example

```python
arr.flags.writeable = False
arr[0] = 100   # ❌ Error
```

---

## 🔹 5. `ALIGNED`

### ✅ Meaning

* Memory is properly aligned for fast access

---

## 🔹 6. `WRITEBACKIFCOPY`

### ✅ Meaning

* Used internally when working with copies

---

# 📘 Example (View vs Copy)

```python
a = np.array([1,2,3])
b = a[1:]   # view

print(b.flags.owndata)
```

👉 Output:

```
False
```

👉 Meaning: `b` does NOT own memory

---

# 📘 Why `flags` is Important?

### ✅ 1. Memory Understanding

Helps debug memory issues

### ✅ 2. Performance

Contiguous arrays → faster

### ✅ 3. Safe Modifications

Check if array is writable

---

# 📘 Quick Summary 🔥

* `flags` → memory info of array
* `C_CONTIGUOUS` → row-wise
* `F_CONTIGUOUS` → column-wise
* `OWNDATA` → owns memory or not
* `WRITEABLE` → can modify or not


#### typecasting
typecasting means converting an array from one data type `(dtype)` to another.

syntax : `arr.astype(dtype, copy=True)`
- dtype -> target data type
- copy:
    - True creates new array (default)
    - False, tries to avoid copy if possible
  
int -> float, float -> int, int -> bool



In [56]:
# int -> float
import numpy as np

arr = np.array([1, 2, 3])
new_arr = arr.astype(float)

print(new_arr)
print(new_arr.dtype)


[1. 2. 3.]
float64


In [58]:
arr = np.array([1.7, 2.9, 3.1])
print(arr.astype(int))

# Values are truncated, not rounded

[1 2 3]


In [59]:
# int -> bool
arr = np.array([0, 1, 2, 0])
print(arr.astype(bool))

[False  True  True False]


In [61]:
# copy behavior
arr = np.array([1,2,3])
new_arr = arr.astype(float) 
# it creates new array 
print(arr is new_arr)  # False

False


In [66]:
# copy = False
another_new_arr = arr.astype(float, copy=False)
another_new_arr


array([1., 2., 3.])

## Structured `dtype` 

- it allows a numpy array to store multiple fields with different data types.
- Store heterogeneous data in one array
Useful for:
    - Tabular data
    - Records (name, age, marks)
    - Binary file handling

Syntax : 

`np.dtype([('field_name1', data_type1),('field_name2', data_type2),
    ...])
`

In [75]:
import numpy as np

dt = np.dtype([
    ('name', 'U10'),   # string (max 10 chars)
    ('age', 'i4'),     # integer (4 bytes)
    ('marks', 'f4')    # float (4 bytes)
])

arr = np.array([
    ('Person1', 21, 85.5),
    ('Person2', 22, 90.0)
], dtype=dt)

print(arr)
print(arr['name']) # accesing the data
print(arr['name'][1])
print(arr[0])
print(arr['marks'][0])

[('Person1', 21, 85.5) ('Person2', 22, 90. )]
['Person1' 'Person2']
Person2
('Person1', 21, 85.5)
85.5


| Code    | Meaning           |
| ------- | ----------------- |
| `'i4'`  | int32 (4 bytes)   |
| `'f4'`  | float32           |
| `'U10'` | string (10 chars) |


### Indexing - single and multidimensional indexing

- for 2D, 3D elements accessing: `arr[row, column]`

In [82]:
arr = np.array([
    [1, 2, 3],
    [4, 5, 6]
])

print(arr[0, 0])   # first row, first column
print(arr[1, 2])   # second row, third column
print(arr[0]) # access entire row
print(arr[:, 1]) # access entire column; 
# : -> means all rows


1
6
[1 2 3]
[2 5]


| Type     | Syntax      | Meaning            |
| -------- | ----------- | ------------------ |
| Basic    | `arr[i]`    | element at index   |
| Negative | `arr[-i]`   | from end           |
| 2D       | `arr[i, j]` | row i, column j    |
| Column   | `arr[:, j]` | all rows, column j |
| Row      | `arr[i, :]` | row i              |



# 📘 1. 1D Slicing

## ✅ Definition

Extract a **subarray from a 1D array** using a range of indices

---

## ✅ Syntax

```python
arr[start : stop : step]
```

### ✅ Parameters

* `start` → starting index (inclusive)
* `stop` → ending index (exclusive)
* `step` → jump size (optional)

---

## ✅ Example

```python
import numpy as np

arr = np.array([10, 20, 30, 40, 50])

print(arr[1:4])
```

👉 Output:

```
[20 30 40]
```

---

## 🔹 Default Behavior

```python
arr[:3]   # from start
arr[2:]   # till end
arr[:]    # full array
```

---

# 📘 2. 2D Slicing

## ✅ Definition

Extract subarrays from **rows and columns**

---

## ✅ Syntax

```python
arr[row_start:row_end, col_start:col_end]
```

---

## ✅ Example

```python
arr = np.array([
    [1,2,3],
    [4,5,6],
    [7,8,9]
])

print(arr[0:2, 1:3])
```

👉 Output:

```
[[2 3]
 [5 6]]
```

👉 Meaning:

* Rows → 0 to 1
* Columns → 1 to 2

---

## 🔹 Full Row / Column

```python
arr[:, 1]   # column
arr[1, :]   # row
```

---

# 📘 3. Step Slicing

## ✅ Definition

Skip elements using `step`

---

## ✅ Example (1D)

```python
arr = np.array([10,20,30,40,50,60])

print(arr[::2])
```

👉 Output:

```
[10 30 50]
```

👉 Meaning: take every 2nd element

---

## 🔹 Reverse Array

```python
print(arr[::-1])
```

👉 Output:

```
[60 50 40 30 20 10]
```

---

## 🔹 2D Step Slicing

```python
print(arr2[:, ::2])
```

👉 Select every 2nd column

---

# 📘 4. View Behavior (VERY IMPORTANT 🔥)

## ✅ Key Concept

👉 **Slicing returns a VIEW, not a copy**

---

## 🔹 Example

```python
arr = np.array([10,20,30,40])

slice_arr = arr[1:3]
slice_arr[0] = 999

print(arr)
```

👉 Output:

```
[10 999 30 40]
```

---

### 🔥 Meaning

* Changing slice → changes original array
* Because both share same memory

---

## 🔹 Check with `base`

```python
print(slice_arr.base is arr)
```

👉 Output:

```
True
```

---

## 🔹 Copy to Avoid This

```python
slice_arr = arr[1:3].copy()
```

👉 Now changes won’t affect original

---

# 📘 Why View is Important?

### ✅ 1. Memory Efficient

* No extra memory used

### ✅ 2. Faster Operations

* No data duplication

---

# 📘 Summary Table 🔥

| Concept           | Meaning          |
| ----------------- | ---------------- |
| `arr[start:stop]` | basic slicing    |
| `arr[:, :]`       | 2D slicing       |
| `step`            | skip elements    |
| `[::-1]`          | reverse          |
| slicing           | returns **view** |

---


### np.empty

In [38]:
np.empty((2,2))
# it creates array without initializing values 
# it will take the random garbage values

array([[0.2, 0.4],
       [0.6, 0.8]])

In [27]:
n = 5
np.identity(n) == np.eye(n) 
# True only when square and default diagonal

array([[ True,  True,  True,  True,  True],
       [ True,  True,  True,  True,  True],
       [ True,  True,  True,  True,  True],
       [ True,  True,  True,  True,  True],
       [ True,  True,  True,  True,  True]])

### np.random.rand

In [39]:
# np,random.rand(d0,d1,...dn)
# geneartes random numbers between 0 and 1

np.random.rand(1,2,3)

array([[[0.86648481, 0.05018639, 0.3322935 ],
        [0.73628003, 0.74415517, 0.81947117]]])

### np.random.randint()


In [40]:
# np.random.randint(low,high=None,size=None,dtype=int)
# size means shape of the output
np.random.randint(10)


1

In [41]:
np.random.randint(100,200,10) # size between 100 and 200 and values are total 10

array([168, 170, 148, 100, 134, 155, 143, 192, 194, 144], dtype=int32)

### Mathmatical operations


a + b
a - b
a * b
a / b
a % b
a ** b

all are works like Matrix addition, sub...like that.


### Broadcasting
- allows operations with different shapes
- shapes must match or
- one dimension must be 1

In [42]:
a = np.array([1,2,3])
print(a + 5)


[6 7 8]


In [43]:
# 2D Broadcasting
a = np.array([[1,2,3],
              [4,5,6]])

b = np.array([10,20,30])

print(a + b)

[[11 22 33]
 [14 25 36]]


In [44]:
np.sqrt(10)


np.float64(3.1622776601683795)

In [45]:

np.square(20)

np.int64(400)

In [46]:
np.abs(-34.5)

np.float64(34.5)

In [50]:
Two_d_array

array([[1, 2, 3],
       [4, 5, 6],
       [7, 8, 9]])

In [56]:
print(np.sum(Two_d_array,axis = 0))
# axis = 0 means column wise
# axis = 1 means row wise

[12 15 18]


### Matrix Multiplication
- np.dot(a,b) dot product
- np.matmul(a,b) or a @ b matrix multiplication
- arr.T matrix transpose

### Comparision Operations

In [63]:
a

array([[1, 2, 3],
       [4, 5, 6]])

In [65]:
a[a==5] = 10
a # it will assisgn val if condition True

array([[ 1,  2,  3],
       [ 4, 10,  6]])

In [66]:
print(a > 4) # check if the val is greater than the 4 if yes it will print True or False

[[False False False]
 [False  True  True]]


In [68]:
print(a[a>4]) # it will print val if val greater than the given element
# a [ a > 4] print val 
# print(a>4) print True or False

[10  6]


### Where condition

In [69]:
# where(condition, True, False)
np.where(a > 2, 100, 0)

array([[  0,   0, 100],
       [100, 100, 100]])

### Reshaping and Manipulation

#### arr.reshape(new_shape) - change the shape
- changing the array shape without changing the data
- Total elements must remain same
- Rules :
- - if there are 6 elements -> 2x3, 3x2, 6x1
  - 4x2 matrix for that it is invalid, elements mismatch
  - arr.reshape(2,-1) Numpy automatically calculates

In [72]:
curr_shape = np.array([1,2,3,4,5,6,7,8,9])
print(curr_shape)

[1 2 3 4 5 6 7 8 9]


In [74]:
new_shape = curr_shape.reshape(3,3)
print(new_shape)

[[1 2 3]
 [4 5 6]
 [7 8 9]]


In [76]:
print(curr_shape.reshape(3,2)) # givese error

ValueError: cannot reshape array of size 9 into shape (3,2)

In [90]:
print(curr_shape.reshape(3,-1)) # auto cal

[[1 2 3]
 [4 5 6]
 [7 8 9]]


#### arr.flatten() and arr.ravel()
- arr.flatten() is converts multi-D array into 1D array & returns a copy
- arr.revel() is converts to 1D array and return view (faster)
- main diff is flatten gives copy and ravel gives view

In [98]:
curr_shape.flatten() # copy

array([1, 2, 3, 4, 5, 6, 7, 8, 9])

In [99]:
curr_shape.ravel() # view

array([1, 2, 3, 4, 5, 6, 7, 8, 9])

#### transpose() or .T
- swaps Rows <--> Columns
- Matrix Operation

In [107]:
mat = np.array([1,2,3,4,5,6,7,8,9]).reshape(3,3)
mat.T # we can use np.transpose(mat)

array([[1, 4, 7],
       [2, 5, 8],
       [3, 6, 9]])

#### Swapaxis & moveaxis

- swapaxes : swap two specific axis(only two). row and col
- syntax : `np.swapaxes(arr, axis1, axis2)`

- moveaxes : move axis to new postion.
- syntax : `np.moveaxis(arr, source, destination)`


In [10]:
import numpy as np
arr = np.array([[1,2,3],[4,5,6],[7,8,9]])
print(arr)
new_arr = np.swapaxes(arr, 0, 1)
print()
print(new_arr)
print(new_arr.shape) 

[[1 2 3]
 [4 5 6]
 [7 8 9]]

[[1 4 7]
 [2 5 8]
 [3 6 9]]
(3, 3)


In [12]:
print(arr)
new_a = np.moveaxis(arr,0,-1)
print(new_a)

[[1 2 3]
 [4 5 6]
 [7 8 9]]
[[1 4 7]
 [2 5 8]
 [3 6 9]]


#### stack & vstack & hstack & dstack & column_stack

| Function       | Key Idea                           |
| -------------- | ---------------------------------- |
| `stack`        | Adds a **new axis**                |
| `vstack`       | Stack **row-wise (vertical)**      |
| `hstack`       | Stack **column-wise (horizontal)** |
| `dstack`       | Stack in **depth (3rd dimension)** |
| `column_stack` | Convert 1D arrays into **columns** |

#### stack
- adds a new dimension(axis)
- `np.stack(arrays, axis=0)`
- axis -> where to insert new axis

#### vstack
- stacks arrays vertically (row-wise)
- `np.vstack(arrays)`

#### hstack
- stacks arrays horizontally (column-wise)
- `np.hstack(arrays)`

#### dstack
- stacks arrays along depth (3D)
- `np.dstack(arrays)`

#### column_stack
- stacks 1D arrays as columns (2D output)

| Function       | Axis     | Dimension change | Output shape (a,b = (3,)) |
| -------------- | -------- | ---------------- | ------------------------- |
| `stack`        | new axis | ✔ adds dimension | `(2,3)` or `(3,2)`        |
| `vstack`       | axis=0   | ❌ no new axis    | `(2,3)`                   |
| `hstack`       | axis=1   | ❌ no new axis    | `(6,)`                    |
| `dstack`       | axis=2   | ✔ makes 3D       | `(1,3,2)`                 |
| `column_stack` | columns  | ❌                | `(3,2)`                   |



In [16]:
# examples for stack
import numpy as np

a = np.array([1,2,3])
b = np.array([4,5,6])
print(a.shape)
print(b.shape)
print(np.stack((a,b))) # changes the shape
print(np.stack((a,b), axis=1)) #with row


(3,)
(3,)
[[1 2 3]
 [4 5 6]]
[[1 4]
 [2 5]
 [3 6]]


In [17]:
import numpy as np

a = np.array([1,2,3])
b = np.array([4,5,6])

res = np.stack((a,b), axis=0)
print(res)
print(res.shape)

[[1 2 3]
 [4 5 6]]
(2, 3)


In [19]:
# example for vstack
a = np.array([1,2,3])
b = np.array([4,5,6])

print(np.vstack((a,b))) # vstack = stack rows (axis=0)
print((np.vstack((a,b))).shape)

[[1 2 3]
 [4 5 6]]
(2, 3)


In [28]:
# example for hstack
a = np.array([1,2,3])
b = np.array([4,5,6])

print(np.stack((a,b),axis = 0))
print(np.hstack((a,b))) # joins arrays side by side
print(np.hstack((a,b)).shape)

[[1 2 3]
 [4 5 6]]
[1 2 3 4 5 6]
(6,)


In [30]:
# example for dstack
a = np.array([1,2,3])
b = np.array([4,5,6])

print(np.dstack((a,b))) # create 3D array each becomes depth


[[[1 4]
  [2 5]
  [3 6]]]


In [32]:
# example for column_stack
# stacks 1D arrays as columns
a = np.array([1,2,3])
b = np.array([4,5,6])

print(np.column_stack((a,b)))
# column_stack = best for tabular data
# Converts 1D → column vectors
# Then stacks side-by-side

[[1 4]
 [2 5]
 [3 6]]


Perfect — now let’s master **splitting in NumPy** with
👉 **clear syntax, parameter meaning, intuition, and examples** 👇

---

# 📘 1. `np.split()`

## ✅ Definition

Splits an array into **equal-sized parts**

---

## ✅ Syntax

```python
np.split(arr, indices_or_sections, axis=0)
```

## ✅ Parameters

* `arr` → input array
* `indices_or_sections` →

  * **int** → number of equal parts
  * **list** → split positions
* `axis` → along which axis to split

---

## 🔹 Example (Equal split)

```python
import numpy as np

arr = np.array([1,2,3,4])

print(np.split(arr, 2))
```

👉 Output:

```
[array([1, 2]), array([3, 4])]
```

---

## 🔹 Example (Using indices)

```python
print(np.split(arr, [1,3]))
```

👉 Output:

```
[array([1]), array([2,3]), array([4])]
```

---

## ⚠️ Important

❌ Must divide equally if using integer

```python
np.split(arr, 3)  # ❌ ERROR
```

---

# 📘 2. `np.array_split()`

## ✅ Definition

Splits array into parts **even if not equal**

---

## ✅ Syntax

```python
np.array_split(arr, sections, axis=0)
```

---

## 🔹 Example

```python
arr = np.array([1,2,3,4])

print(np.array_split(arr, 3))
```

👉 Output:

```
[array([1,2]), array([3]), array([4])]
```

---

## 🧠 Insight

👉 Unequal distribution allowed

---

# 📘 3. `np.vsplit()`

## ✅ Definition

Splits array **vertically (row-wise)**

---

## ✅ Syntax

```python
np.vsplit(arr, sections)
```

👉 Equivalent to:

```python
np.split(arr, sections, axis=0)
```

---

## 🔹 Example

```python
arr = np.array([[1,2],
                [3,4]])

print(np.vsplit(arr, 2))
```

👉 Output:

```
[array([[1, 2]]), array([[3, 4]])]
```

---

## 🧠 Insight

👉 Splits into rows

---

# 📘 4. `np.hsplit()`

## ✅ Definition

Splits array **horizontally (column-wise)**

---

## ✅ Syntax

```python
np.hsplit(arr, sections)
```

👉 Equivalent to:

```python
np.split(arr, sections, axis=1)
```

---

## 🔹 Example

```python
arr = np.array([[1,2],
                [3,4]])

print(np.hsplit(arr, 2))
```

👉 Output:

```
[array([[1],
        [3]]), 
 array([[2],
        [4]])]
```

---

## 🧠 Insight

👉 Splits into columns

---

# 📘 5. `np.dsplit()`

## ✅ Definition

Splits array along **depth (3rd axis)**

---

## ✅ Syntax

```python
np.dsplit(arr, sections)
```

👉 Equivalent to:

```python
np.split(arr, sections, axis=2)
```

---

## 🔹 Example

```python
arr = np.ones((2,2,2))

print(np.dsplit(arr, 2))
```

👉 Output:

```
[array([[[1.],
         [1.]],
        [[1.],
         [1.]]]),
 array([[[1.],
         [1.]],
        [[1.],
         [1.]]])]
```

---

## 🧠 Insight

👉 Works only for **3D arrays**

---

# 📘 🔥 BIG INTUITION

## 👉 Think of axes like directions:

| Axis     | Direction            |
| -------- | -------------------- |
| `axis=0` | vertical (rows)      |
| `axis=1` | horizontal (columns) |
| `axis=2` | depth (layers)       |

---

# 📘 🔥 Comparison Table

| Function      | Axis   | Equal required? | Use case         |
| ------------- | ------ | --------------- | ---------------- |
| `split`       | any    | ✔ Yes           | strict splitting |
| `array_split` | any    | ❌ No            | flexible         |
| `vsplit`      | axis=0 | ✔               | row split        |
| `hsplit`      | axis=1 | ✔               | column split     |
| `dsplit`      | axis=2 | ✔               | depth split      |

---

# 📘 🔥 Shape Understanding Example

```python
arr = np.array([[1,2,3],
                [4,5,6]])
```

### `vsplit(arr, 2)`

```
→ 2 arrays of shape (1,3)
```

### `hsplit(arr, 3)`

```
→ 3 arrays of shape (2,1)
```

---

# 📘 🚨 Common Mistakes

### ❌ Using `split()` when unequal

✔ Use `array_split()`

---

### ❌ Confusing vsplit vs hsplit

✔ `vsplit` → rows
✔ `hsplit` → columns

---

### ❌ Using dsplit on 2D

✔ Only works on 3D

---

# 🧠 Interview One-Liner

👉
**NumPy splitting functions divide arrays along specified axes, where `split` requires equal division, `array_split` allows flexibility, and `vsplit`, `hsplit`, `dsplit` are axis-specific shortcuts.**

---


#### np.concatenate((a,b), axis = 0)
- join the array
- in axis, 0 means row-wise vertically, 1 means Column-wise horizontally

In [108]:
a = np.array([[1,2]])
b = np.array([[3,4]])
np.concatenate((a,b),axis = 0)

array([[1, 2],
       [3, 4]])

In [109]:
np.concatenate((a,b), axis = 1)

array([[1, 2, 3, 4]])

#### Handling missing values
- np.nan -> not a number used most for floats
- None -> Python object (but less preferred)
- np.nan represents missing/undefined numeric value

In [110]:
demo = np.array([1,2,np.nan])
print(demo)

[ 1.  2. nan]


In [112]:
np.isnan(demo) # return True if nan exists otherwise False

array([False, False,  True])

In [115]:
demo[~np.isnan(demo)] # it removes the nan values

array([1., 2.])

In [119]:
# Replace nan or missing values
np.nan_to_num(demo,nan=3)

# np.nanmean() -> mean ignoring NaN
# np.nansum() -> sum ignoring NaN
# np.nanmin()
# np.nanmax()


array([1., 2., 3.])

In [122]:
np.nan == np.nan # value is unknown unknown == unknown X 

False

In [123]:
np.isnan(demo)

array([False, False,  True])

In [124]:
arr = np.array([[1, np.nan],
                [3, 4]])
arr[~np.isnan(arr).any(axis=1)]

array([[3., 4.]])

#### iteration in numpy

- normally multi dimensional iteration is quite difficult in numpy.2D or 3D arrays you get rows not elements while you iterating....!!
- To iterate element by element, numpy provides :
    - `nditer`
    - `flat iterator`
#### nditer

- good & efficient for n-dimensional array
- `np.nditer(arr,flags=None,op_flags=None`
- flags -> to control iteration behaviour
    - 'c_index' - track index
    - 'multi_index' - multi-dim index
- op_flags -> to control `readonly` & `readwrite` allow modifications

#### arr.flat

- simple iterator that treats array as 1D
- `arr.flat`



In [4]:
import numpy as np

arr = np.array([[1,2],
                [3,4]])

for x in np.nditer(arr):
    print(x)

1
2
3
4


In [5]:
for x in np.nditer(arr, op_flags=['readwrite']): # if readonly we cannot change
    x[...] = x * 2

print(arr)

[[2 4]
 [6 8]]


In [10]:
for x in np.nditer(arr, flags=['multi_index']):
    print(x, "Index:", x) # helps when position matters

2 Index: 2
4 Index: 4
6 Index: 6
8 Index: 8


In [11]:
for x in np.nditer(arr, flags=['external_loop']):
    print(x)

[2 4 6 8]


In [13]:
# arr.flat
arr = np.array([[1,2],
                [3,4]])

for x in arr.flat:
    print(x) # like flatten

1
2
3
4


#### Value Transformation functions
`np.clip(arr, a_min, a_max)`
- limits values within the range
- it will takes from a_min, a_max


In [14]:
import numpy as np

arr = np.array([1, 5, 10, 15])
print(arr)
print(np.clip(arr, 3, 12))

[ 1  5 10 15]
[ 3  5 10 12]


In [15]:
np.round(arr, decimals=0) # round to nearest value

array([ 1,  5, 10, 15])

In [16]:
arr = np.array([1.2, 2.5, 3.7])

print(np.round(arr))

[1. 2. 4.]


In [17]:
np.round(arr, 1)

array([1.2, 2.5, 3.7])

In [18]:
np.floor([1.7, 2.3, -1.2]) # moves backward -OO round down to nearest value


array([ 1.,  2., -2.])

In [20]:
np.ceil([1.2, 2.3, -1.7]) # rounds up to nearest integer

array([ 2.,  3., -1.])

In [22]:
np.trunc([1.7, -1.7]) # removes decimal part (no rounding)

array([ 1., -1.])



# 📘 1. `np.ix_()` — Think: “Select rows AND columns properly”

---

## 🔴 First understand the problem

```python
import numpy as np

arr = np.array([[10,20,30],
                [40,50,60],
                [70,80,90]])

rows = [0,2]
cols = [0,1]

print(arr[rows, cols])
```

👉 Output:

```text
[10 80]
```

---

## ❓ Why this output?

It pairs:

* (0,0) → 10
* (2,1) → 80

👉 ❌ It does NOT give full submatrix

---

# ✅ Now use `np.ix_()`

```python
print(arr[np.ix_(rows, cols)])
```

👉 Output:

```text
[[10 20]
 [70 80]]
```

---

## 🔍 What happened?

It created all combinations:

```text
(0,0) (0,1)
(2,0) (2,1)
```

---

## 🧠 Simple Idea

👉 `ix_` = **give me all combinations of rows & columns**

---

## 🎯 Real meaning

👉 “Give me rows 0 & 2 AND columns 0 & 1”

---

---

# 📘 2. `np.take()` — Think: “Pick items by position”

---

## 🔹 Example (1D)

```python
arr = np.array([10,20,30,40])

print(np.take(arr, [0,2]))
```

👉 Output:

```text
[10 30]
```

---

## 🧠 Meaning

👉 “Take index 0 and 2”

---

---

## 🔹 Example (2D)

```python
arr = np.array([[10,20,30],
                [40,50,60]])

print(np.take(arr, [0,2], axis=1))
```

👉 Output:

```text
[[10 30]
 [40 60]]
```

---

## 🔍 What happened?

👉 axis=1 → columns

👉 took column 0 and 2

---

## 🧠 Simple Idea

👉 `take` = **pick elements along a direction**

---

---

# 📘 3. `np.put()` — Think: “Replace items”

---

## 🔹 Example (1D)

```python
arr = np.array([10,20,30,40])

np.put(arr, [1,3], [99,88])

print(arr)
```

👉 Output:

```text
[10 99 30 88]
```

---

## 🧠 Meaning

👉 “Replace index 1 → 99, index 3 → 88”

---

---

## 🔹 Important (VERY IMPORTANT ⚠️)

👉 `put` works on **flattened array**

---

## 🔹 Example (2D)

```python
arr = np.array([[1,2],
                [3,4]])

np.put(arr, [0,3], [9,8])

print(arr)
```

👉 Output:

```text
[[9 2]
 [3 8]]
```

---

## 🔍 Why?

Flattened array:

```text
[1, 2, 3, 4]
```

👉 index 0 → 1 → 9
👉 index 3 → 4 → 8

---

## 🧠 Simple Idea

👉 `put` = **replace values using flat index**

---

---

# 📘 🔥 Final Intuition (SUPER IMPORTANT)

---

## 👉 `ix_`

Think:

```text
“I want a submatrix (rows + columns combination)”
```

---

## 👉 `take`

Think:

```text
“I want to pick elements by index”
```

---

## 👉 `put`

Think:

```text
“I want to replace elements at positions”
```

---

---

# 📘 🔥 Visual Summary

| Function | Think Like     |
| -------- | -------------- |
| `ix_`    | select grid    |
| `take`   | pick values    |
| `put`    | replace values |

---

---

# 📘 🚨 Common Mistake (VERY IMPORTANT)

---

### ❌ This confusion

```python
arr[rows, cols]
```

👉 Gives pair-wise selection

---

### ✔ Correct way

```python
arr[np.ix_(rows, cols)]
```

👉 Gives full submatrix

---

---

# 🧠 FINAL SIMPLE EXPLANATION

👉

* **`ix_`** → gives combinations (rows × columns)
* **`take`** → extracts elements by index
* **`put`** → replaces elements using index

---


Great — this is a **very important real-world NumPy topic** 🔥
Handling missing values (`NaN`) is asked frequently in interviews.

I’ll explain clearly with:
✔ Concept
✔ Syntax + parameters
✔ Examples
✔ Interview insights

---

# 📘 1. What is `NaN`?

## ✅ Definition

`NaN` = **Not a Number**

👉 Represents **missing / undefined values**

---

## 🔹 Example

```python id="1s1u36"
import numpy as np

arr = np.array([1, 2, np.nan, 4])
print(arr)
```

👉 Output:

```id="qv9j74"
[ 1.  2. nan  4.]
```

---

## 🧠 Important Properties

### ❗ NaN is NOT equal to anything

```python id="c1dskl"
print(np.nan == np.nan)
```

👉 Output:

```id="j4t3po"
False
```

---

### ✔ Use special functions instead

---

# 📘 2. `np.isnan()`

## ✅ Definition

Checks whether elements are `NaN`

---

## ✅ Syntax

```python id="ifh2zb"
np.isnan(arr)
```

---

## 🔹 Example

```python id="qf6k3c"
arr = np.array([1, np.nan, 3])

print(np.isnan(arr))
```

👉 Output:

```id="qkm1u8"
[False  True False]
```

---

## 🧠 Insight

👉 Returns **boolean mask**

---

# 📘 3. Problem with Normal Functions ⚠️

```python id="m6a6bt"
arr = np.array([1,2,np.nan,4])

print(np.mean(arr))
```

👉 Output:

```id="g1e3c6"
nan
```

👉 ❌ NaN spreads → result becomes NaN

---

# 📘 4. NaN-Aware Functions 🔥

## ✅ Definition

Functions that **ignore NaN values**

---

---

# 📘 4.1 `np.nanmean()`

## ✅ Syntax

```python id="1h2c3z"
np.nanmean(arr, axis=None)
```

---

## 🔹 Example

```python id="7j7cs1"
arr = np.array([1,2,np.nan,4])

print(np.nanmean(arr))
```

👉 Output:

```id="1f6qnp"
2.3333333333333335
```

👉 Calculation:

```text id="oahg4h"
(1 + 2 + 4) / 3
```

---

# 📘 4.2 `np.nanstd()`

## ✅ Syntax

```python id="jjqln2"
np.nanstd(arr, axis=None)
```

---

## 🔹 Example

```python id="rtz8rf"
print(np.nanstd(arr))
```

👉 Ignores NaN while computing standard deviation

---

# 📘 4.3 Other NaN-aware Functions 🔥

| Function       | Meaning               |
| -------------- | --------------------- |
| `np.nanmean()` | mean ignoring NaN     |
| `np.nanstd()`  | std ignoring NaN      |
| `np.nanvar()`  | variance ignoring NaN |
| `np.nanmin()`  | min ignoring NaN      |
| `np.nanmax()`  | max ignoring NaN      |
| `np.nansum()`  | sum ignoring NaN      |

---

## 🔹 Example

```python id="vnlx9z"
print(np.nansum(arr))
```

👉 Output:

```id="9rpk6o"
7.0
```

---

# 📘 5. Replace NaN Values

## 🔹 Using `np.nan_to_num()`

```python id="z6q44g"
arr = np.array([1, np.nan, 3])

print(np.nan_to_num(arr, nan=0))
```

👉 Output:

```id="pkk0cv"
[1. 0. 3.]
```

---

# 📘 6. Filtering NaN Values

```python id="55cbsc"
arr = np.array([1, np.nan, 3])

clean = arr[~np.isnan(arr)]
print(clean)
```

👉 Output:

```id="xk3hlz"
[1. 3.]
```

---

# 📘 7. Axis with NaN Functions

```python id="5w1z5z"
arr = np.array([[1,np.nan],
                [3,4]])

print(np.nanmean(arr, axis=0))
```

👉 Output:

```id="y8pxx1"
[2. 4.]
```

---

# 📘 🔥 Real-World Example

```python id="j4c2kg"
marks = np.array([80, 90, np.nan, 70])

avg = np.nanmean(marks)
print("Average:", avg)
```

👉 Missing marks ignored

---

# 📘 🔥 Key Differences

| Function    | Behavior       |
| ----------- | -------------- |
| `mean()`    | fails with NaN |
| `nanmean()` | ignores NaN    |

---

# 📘 🚨 Common Mistakes

### ❌ Using `== np.nan`

✔ Always use:

```python id="rb9ztu"
np.isnan(arr)
```

---

### ❌ Using normal functions

✔ Use:

```python id="yd6e6t"
np.nanmean()
```

---

# 🧠 Interview One-Liner

👉
**NumPy provides NaN-aware functions like `nanmean` and `nanstd` that ignore missing values, enabling robust statistical computations on incomplete datasets.**

---



### Linear algebra

- used for matrix operations
- solving equations
- ML Algorithms
- Transformations
#### np.linalg

In [134]:
# Inverse of matrix
Two_d_array
x = np.linalg.inv(Two_d_array)
print(x)
A = np.array([[1,2],[3,4]])

# Try these:
np.linalg.det(A)
np.linalg.inv(A)
np.linalg.eig(A)
np.trace(A)

[[ 3.15251974e+15 -6.30503948e+15  3.15251974e+15]
 [-6.30503948e+15  1.26100790e+16 -6.30503948e+15]
 [ 3.15251974e+15 -6.30503948e+15  3.15251974e+15]]


np.int64(5)

### Views and Copies

- *View* -> Shares same Memory
- *Copy* -> Creates new Memory
#### View
- A view does not create new data it just references the original array.
- syntax : ` arr.view()`
- so it is faster (no new memory)
- used for performance optimization
#### Copy
- Creates a completely new array
- syntax : `arr.copy()`
- orginal data unaffected
- Needed when modifying data
- `arr.base` to check view or copy
- if base != None -> View
- if base = None -> Copy

Lets goto `flatten() & ravel()`
flatten is returns copy
ravel is returns view

In [141]:
data = np.array([1,2,3,4,5])
data2 = data.copy()
data2[4] = 100
print(data) # not affected
print(data.base)
print(data2.base) # None means Copy

[1 2 3 4 5]
None
None


In [147]:
data3 = data.view()
data4 = data
data4[1]=100
print(data)
print(data2)
print(data3) # data modified
print(data4)
print(data3.base) # no None so it is View

[  1 100   3   4   5]
[  1   2   3   4 100]
[  1 100   3   4   5]
[  1 100   3   4   5]
[  1 100   3   4   5]


### Sorting and Searching in Numpy

`np.sort(arr,axis = -1, kind='quicksort, order = None)`
- - returns a sorted copy of array
- - Does't change the original array
- - axis -> which direction sorting happens
- - - axis = 0 sort column wise
- - - axis = 1 sort row wise
- - - axis = -1 last axis by default
- - - axis = None -> flatten + sort converts to 1D then sorts
- - kind -> sorting algorithm



In [153]:
arr = np.array([[3,12,1],
                [6,5,14]])

np.sort(arr, axis=1)

array([[ 1,  3, 12],
       [ 5,  6, 14]])

In [154]:
np.sort(arr,axis=0)

array([[ 3,  5,  1],
       [ 6, 12, 14]])

In [155]:
np.sort(arr, axis=None)

array([ 1,  3,  5,  6, 12, 14])

In [157]:
arr.sort(axis=-1, kind='quicksort')
arr

array([[ 1,  3, 12],
       [ 5,  6, 14]])

In [163]:
arr = np.array([[3,2,1],
                [6,5,4]])

np.argsort(arr) # return indices that sort array
# np.argsort(arr,axis=-1,kind='quicksort')
# sorting based on indices then it is useful

array([[2, 1, 0],
       [2, 1, 0]])

In [164]:
arr = np.array([30,10,20])
np.argsort(arr)

array([1, 2, 0])

#### Searching

#### np.searchsorted()
`np.searchsorted(a, v, side='left', sorter=None)`

a - sorted array

v - value to insert

side - left or right

sorter = optional index array


In [167]:
arr = np.array([1,3,5,7])
np.searchsorted(arr, 5

np.int64(2)

In [168]:
# side differences
np.searchsorted(arr, 5, side='left')   # 2
np.searchsorted(arr, 5, side='right')  # 3

np.int64(3)

#### np.unique()

`np.unique(a, return_index=False, return_counts=False)`

removes the duplicates


In [169]:
np.unique([1,2,2,3])
np.unique(arr, return_counts=True)

(array([1, 3, 5, 7]), array([1, 1, 1, 1]))

In [166]:
# np.where(condition, True, False)
arr = np.array([1,2,3,4])

np.where(arr > 2)

(array([2, 3]),)

## File handling in Numpy

Here is a **complete, clean, copyable Markdown guide** for **File Handling in NumPy** with **syntax, parameters, explanations, and examples** 👇

---

# 📂 NumPy File Handling (Complete Guide)

NumPy provides powerful functions to **save and load arrays** in different formats:

* Text files (`.txt`, `.csv`)
* Binary files (`.npy`, `.npz`)

---

# 🔹 1. Saving Files

---

## ✅ 1.1 `np.save()` → Save as Binary (.npy)

### 🔹 Syntax

```python
np.save(file, arr, allow_pickle=True, fix_imports=True)
```

### 🔹 Parameters

| Parameter      | Meaning                           |
| -------------- | --------------------------------- |
| `file`         | File name (string or file object) |
| `arr`          | NumPy array to save               |
| `allow_pickle` | Allows saving Python objects      |
| `fix_imports`  | Compatibility for Python 2        |

### 🔹 Example

```python
import numpy as np

arr = np.array([1, 2, 3, 4])

np.save("data.npy", arr)
```

---

## ✅ 1.2 `np.savez()` → Save Multiple Arrays (.npz)

### 🔹 Syntax

```python
np.savez(file, *args, **kwargs)
```

### 🔹 Parameters

| Parameter  | Meaning                           |
| ---------- | --------------------------------- |
| `file`     | File name                         |
| `*args`    | Arrays (saved as arr_0, arr_1...) |
| `**kwargs` | Named arrays                      |

### 🔹 Example

```python
a = np.array([1, 2])
b = np.array([3, 4])

np.savez("data.npz", a=a, b=b)
```

---

## ✅ 1.3 `np.savetxt()` → Save as Text (.txt / .csv)

### 🔹 Syntax

```python
np.savetxt(fname, X, fmt='%.18e', delimiter=' ', newline='\n',
           header='', footer='', comments='# ', encoding=None)
```

### 🔹 Parameters

| Parameter   | Meaning                    |
| ----------- | -------------------------- |
| `fname`     | File name                  |
| `X`         | Array data                 |
| `fmt`       | Format (e.g. `%d`, `%.2f`) |
| `delimiter` | Separator (`,` for CSV)    |
| `newline`   | Line separator             |
| `header`    | Header text                |
| `footer`    | Footer text                |
| `comments`  | Prefix for header/footer   |
| `encoding`  | File encoding              |

### 🔹 Example

```python
arr = np.array([[1, 2, 3], [4, 5, 6]])

np.savetxt("data.txt", arr, fmt='%d', delimiter=',')
```

---

# 🔹 2. Loading Files

---

## ✅ 2.1 `np.load()` → Load Binary (.npy / .npz)

### 🔹 Syntax

```python
np.load(file, mmap_mode=None, allow_pickle=False, fix_imports=True)
```

### 🔹 Parameters

| Parameter      | Meaning               |
| -------------- | --------------------- |
| `file`         | File name             |
| `mmap_mode`    | Memory mapping mode   |
| `allow_pickle` | Allow loading objects |
| `fix_imports`  | Python compatibility  |

### 🔹 Example (npy)

```python
data = np.load("data.npy")
print(data)
```

### 🔹 Example (npz)

```python
data = np.load("data.npz")
print(data['a'])
print(data['b'])
```

---

## ✅ 2.2 `np.loadtxt()` → Load Text Files

### 🔹 Syntax

```python
np.loadtxt(fname, dtype=float, comments='#', delimiter=None,
           skiprows=0, usecols=None, unpack=False)
```

### 🔹 Parameters

| Parameter   | Meaning              |
| ----------- | -------------------- |
| `fname`     | File name            |
| `dtype`     | Data type            |
| `comments`  | Ignore comment lines |
| `delimiter` | Separator            |
| `skiprows`  | Skip rows            |
| `usecols`   | Select columns       |
| `unpack`    | Transpose output     |

### 🔹 Example

```python
data = np.loadtxt("data.txt", delimiter=',')
print(data)
```

---

## ✅ 2.3 `np.genfromtxt()` → Advanced Loading (Handles Missing Data)

### 🔹 Syntax

```python
np.genfromtxt(fname, dtype=float, delimiter=',',
              skip_header=0, skip_footer=0,
              missing_values=None, filling_values=None)
```

### 🔹 Parameters

| Parameter        | Meaning                 |
| ---------------- | ----------------------- |
| `fname`          | File name               |
| `dtype`          | Data type               |
| `delimiter`      | Separator               |
| `skip_header`    | Skip top rows           |
| `skip_footer`    | Skip bottom rows        |
| `missing_values` | Missing value indicator |
| `filling_values` | Replace missing values  |

### 🔹 Example

```python
data = np.genfromtxt("data.txt", delimiter=',', filling_values=0)
print(data)
```

---

# 🔹 3. File Types Summary

| Format | Function       | Description                 |
| ------ | -------------- | --------------------------- |
| `.npy` | `np.save()`    | Single array (binary, fast) |
| `.npz` | `np.savez()`   | Multiple arrays             |
| `.txt` | `np.savetxt()` | Human-readable              |
| `.csv` | `np.savetxt()` | Comma-separated             |

---

# 🔹 4. Important Notes

### ⚠️ Binary vs Text

* `.npy` / `.npz` → Fast, compact
* `.txt` / `.csv` → Readable but slower

---

### ⚠️ When to Use What

| Use Case        | Recommended    |
| --------------- | -------------- |
| Large data      | `.npy`         |
| Multiple arrays | `.npz`         |
| Sharing data    | `.csv`         |
| Missing values  | `genfromtxt()` |

---

# 🔹 5. Complete Example (End-to-End)

```python
import numpy as np

# Create data
arr = np.array([[1, 2, 3], [4, 5, 6]])

# Save
np.save("data.npy", arr)
np.savetxt("data.csv", arr, delimiter=',', fmt='%d')

# Load
data1 = np.load("data.npy")
data2 = np.loadtxt("data.csv", delimiter=',')

print(data1)
print(data2)
```

---

# 🔹 6. Interview Tips 🔥

* `np.save` → fastest method
* `np.loadtxt` vs `genfromtxt` → missing data handling
* `.npz` → dictionary-like access
* Always specify `delimiter` for CSV
* Use `fmt='%d'` or `'%.2f'` in text saving

---

If you want next, I can give:
✅ **NumPy Searching & Sorting (full like this)**
✅ **NumPy Random module complete**
✅ **Interview questions on NumPy file handling**


# cheatsheet
````markdown
# 🧠 NumPy Complete Cheat Sheet (Interview Ready)

> Covers: Arrays, Indexing, Operations, Broadcasting, Math, Stats, Linear Algebra, Random, Searching, Sorting, File Handling

---

# 🔹 1. Import NumPy
```python
import numpy as np
````

---

# 🔹 2. Creating Arrays

## ✅ Basic Creation

```python
a = np.array([1, 2, 3])
# Output: array([1, 2, 3])
# Why: Converts Python list → NumPy array

b = np.array([[1,2],[3,4]])
# Output: 2D array
```

## ✅ Special Arrays

```python
np.zeros((2,3))
# Output: [[0. 0. 0.]
#          [0. 0. 0.]]
# Why: Initialize with zeros

np.ones((2,2))
# Output: all ones

np.full((2,2), 7)
# Output: [[7 7],[7 7]]

np.eye(3)
# Output: Identity matrix

np.arange(0,10,2)
# Output: [0 2 4 6 8]
# Why: start, stop, step

np.linspace(0,1,5)
# Output: 5 evenly spaced values
```

---

# 🔹 3. Array Attributes

```python
a = np.array([[1,2,3],[4,5,6]])

a.ndim   # Output: 2 → dimensions
a.shape  # Output: (2,3) → rows, cols
a.size   # Output: 6 → total elements
a.dtype  # Output: int64 → type
```

---

# 🔹 4. Indexing & Slicing

```python
a = np.array([10,20,30,40])

a[0]      # 10
a[-1]     # 40

a[1:3]    # [20 30]
# Why: slicing like Python list

b = np.array([[1,2],[3,4]])
b[0,1]    # 2
```

---

# 🔹 5. Reshaping

```python
a = np.arange(6)

a.reshape(2,3)
# Output: [[0 1 2],[3 4 5]]
# Why: change shape without changing data

a.flatten()
# Output: 1D array
```

---

# 🔹 6. Data Types

```python
a = np.array([1,2,3], dtype=float)
# Output: [1. 2. 3.]

a.astype(int)
# Convert type
```

---

# 🔹 7. Arithmetic Operations

```python
a = np.array([1,2,3])
b = np.array([4,5,6])

a + b   # [5 7 9]
a - b   # [-3 -3 -3]
a * b   # [4 10 18]
a / b   # division

# Why: Element-wise operations
```

---

# 🔹 8. Broadcasting

```python
a = np.array([1,2,3])
b = 2

a * b
# Output: [2 4 6]
# Why: scalar broadcast to match shape
```

---

# 🔹 9. Aggregation Functions

```python
a = np.array([1,2,3,4])

np.sum(a)   # 10
np.mean(a)  # 2.5
np.max(a)   # 4
np.min(a)   # 1
np.std(a)   # standard deviation

# Axis example
b = np.array([[1,2],[3,4]])

np.sum(b, axis=0)
# Output: [4 6] (column-wise)
```

---

# 🔹 10. Mathematical Functions

```python
np.sqrt(a)   # square root
np.exp(a)    # exponential
np.log(a)    # log
np.sin(a)    # sine
```

---

# 🔹 11. Searching

```python
a = np.array([10,20,30,20])

np.where(a == 20)
# Output: (array([1,3]),)
# Why: indices where condition true

np.argmax(a)
# Output: 2 (index of max)

np.argmin(a)
```
#### np.nonzero()
✅ Definition
- Returns indices of non-zero elements
- Syntax
np.nonzero(arr)
🔹 Example
arr = np.array([0,1,0,2,3])

print(np.nonzero(arr))

(array([1,3,4]),)

---

# 🔹 12. Sorting

```python
a = np.array([3,1,2])

np.sort(a)
# Output: [1 2 3]

np.argsort(a)
# Output: [1 2 0] (indices)

b = np.array([[3,2],[1,4]])

np.sort(b, axis=0)
# column-wise sorting
```

---

# 🔹 13. Filtering

```python
a = np.array([10,20,30,40])

a[a > 20]
# Output: [30 40]
# Why: boolean indexing
```

---

# 🔹 14. Stacking & Splitting

```python
a = np.array([1,2])
b = np.array([3,4])

np.vstack((a,b))
# vertical stack

np.hstack((a,b))
# horizontal stack

np.split(np.array([1,2,3,4]), 2)
# Output: [array([1,2]), array([3,4])]
```

---

# 🔹 15. Copy vs View

```python
a = np.array([1,2,3])
b = a.copy()
c = a.view()

# copy → independent
# view → reflects changes
```

---

# 🔹 16. Linear Algebra

```python
a = np.array([[1,2],[3,4]])

np.transpose(a)
# transpose

np.dot(a,a)
# matrix multiplication

np.linalg.det(a)
# determinant

np.linalg.inv(a)
# inverse
```

---

# 🔹 17. Random Module

```python
np.random.rand(2,2)
# random floats (0–1)

np.random.randint(1,10,(2,2))
# random integers

np.random.choice([1,2,3], size=5)
# random selection

np.random.seed(42)
# reproducible results
```

---

# 🔹 18. File Handling

## Save

```python
np.save("data.npy", a)
# binary save

np.savetxt("data.txt", a, delimiter=',')
# text save
```

## Load

```python
np.load("data.npy")
# load binary

np.loadtxt("data.txt", delimiter=',')
# load text
```

---

# 🔹 19. Iteration

```python
for x in np.nditer(a):
    print(x)
# iterate elements
```

---

# 🔹 20. Important Interview Notes 🚀

* NumPy is faster due to **vectorization**
* Avoid loops → use vector operations
* Broadcasting is key concept
* `axis` is frequently asked
* Difference:

  * `copy()` vs `view()`
  * `loadtxt()` vs `genfromtxt()`
* `.shape` vs `.reshape()`

---

# 🔥 Quick Revision Table

| Concept      | Function  |
| ------------ | --------- |
| Create array | np.array  |
| Range        | np.arange |
| Reshape      | reshape   |
| Sum          | np.sum    |
| Sort         | np.sort   |
| Search       | np.where  |
| Random       | np.random |
| Save         | np.save   |
| Load         | np.load   |

---

# ✅ Final Tip

👉 Practice:

* Array operations
* Broadcasting
* Indexing tricks

👉 Most asked:

* `axis`
* `reshape`
* `where`
* `argsort`

---

```
```


